### PR TITLE
Preserve chronological order of stdout and stderr with ``capsys``

### PR DESCRIPTION
**Pull request by aklajnert**
This is a solution that will allow preserving chronological order of stdout and stderr while capturing (#5449). 

The solution changes `capsys` to use `deque` for keeping the order of streams. It is possible to retrieve both streams traditionally or as a combined version.  

I've added two arguments to `readouterr()` but only for `capsys`. I wonder if that's a good approach, or should I add a new method for retrieving combined output? An additional argument is to not flush streams while retrieving. This will allow retrieving the streams traditionally *and* combined at the same time.

### CODE REVIEW

**Comment by blueyed** [comment]
Thanks for working on this.
I think it might be easier though to just share the same `tmpfile`?
This would be more similar to `2>&1` (to redirect stderr to stdout).
Of course it would not allow for reading them separately anymore, which could be useful.

**Comment by blueyed** [comment]
btw: see also https://github.com/pytest-dev/pytest/pull/6671 (feedback would be appreciated).  This should also be done for `features`, where other code there changed already.

**Comment by aklajnert** [comment]
> btw: see also #6671 (feedback would be appreciated). This should also be done for `features`, where other code there changed already.

Right, I wanted to put it in `features`, but I've forgot to change it. 

In theory, if I'll put tuples into `deque`, the memory footprint of this solution shouldn't be much bigger than the current `capsys`. If that is true, it would make sense to replace `capsys` with it. So basically this PR would just extend the current `capsys` capability with the feature to get both streams in the correct order.

File: src/_pytest/capture.py [hunk 1]
```diff
@@ -451,6 +468,10 @@ class MultiCapture:
     _state = None
 
     def __init__(self, out=True, err=True, in_=True, Capture=None):
+        if Capture == OrderedCapture:
+            self.streams = collections.deque()
+            Capture = functools.partial(OrderedCapture, (self.streams))
```
Comment (blueyed) [inline src/_pytest/capture.py]: This should rather be passed in as a separate `Capture` class, i.e. use `functools.partial` before passing it here.

File: testing/test_capture.py [hunk 2]
```diff
@@ -1510,3 +1510,27 @@ def test_fails():
...
+    assert (
+        out == "stdout1\n"
+        "stdout2\n"
+        "stderr1\n"
+        "stdout3\n"
+        "stderr2\n"
+        "stderr3\n"
+        "stdout4\n"
+        "stdout5\n"
+    )
```
Comment (blueyed) [inline testing/test_capture.py]: Shoudl also test non-combined output (and make the assert more readable.. :))

Comment (aklajnert) [reply testing/test_capture.py]: This test was more a demo than real test, I'll also cover the non-combined use case. What's wrong with the asserts?

File: src/_pytest/capture.py (hunk 1 above)
Comment (aklajnert) [reply src/_pytest/capture.py]: Why do you think that'd be better? The only other place where I can use `fuctools.partial` is in `CaptureFixture` class, but I don't see any added value here.

**Comment by aklajnert** [comment]
I've changed the capture class for `capsys` to the `OrderedCapture`, and removed the `capsysord` fixture. Now the capsys has the capability to combine streams, and there is not much memory overhead. All tests for capture are passing.

File: src/_pytest/capture.py (hunk 1 above)
Comment (blueyed) [reply src/_pytest/capture.py]: I think the logic should not go into MultiCapture really, but should be handled by the passed in class/callable.

File: testing/test_capture.py (hunk 2 above)
Comment (blueyed) [reply testing/test_capture.py]: > What's wrong with the asserts?

It's actually ok.

File: src/_pytest/capture.py (hunk 1 above)
Comment (blueyed) [reply src/_pytest/capture.py]: I've not checked where/if you can do it in another way really though, but here's an example for having a capture class that resumes on reading from stdin (just as an example really, https://github.com/blueyed/pytest/blob/eda98a194a4b512a1fa5aba8e45f9bbe1f1866a1/src/_pytest/capture.py#L98-L111).

Comment (aklajnert) [reply src/_pytest/capture.py]: I've changed the approach to have the `deque` as class attribute of the `OrderedWriter`.  
Anyway, I'll wait for you to merge #6671 before I'll finish this branch.

**Comment by bluetech** [comment]
Hi @aklajnert,

We've recently stopped using the `features` branch, and use the `master` branch instead ([rationale](https://github.com/pytest-dev/pytest/pull/6571)). We're asking open PRs against `features` to be rabased on `master`. You can do it as follows:

- `git checkout master && git pull origin` - sync the latest code.
- `git checkout 5449_preserve_capsys_stream_order` - checkout the branch of this PR.
- `git rebase -i master` - leave only your own commits, and delete everything else.
- `git push --force` - update the PR.
- Click `Edit` in the PR GitHub UI and change the base branch from `features` to `master`.

Let us know if you have any issues.

Thanks!

File: src/_pytest/capture.py [hunk 3]
```diff
@@ -692,6 +704,39 @@ def snap(self):
         return res
 
 
+class OrderedCapture(SysCapture):
+    """This class uses deque to keep streams order."""
```
Comment (blueyed) [inline src/_pytest/capture.py]:
```suggestion
    """This class uses deque to keep streams in order."""
```

File: src/_pytest/capture.py [hunk 4]
```diff
@@ -692,6 +704,39 @@ def snap(self):
...
+        if self.name == "stderr":
+            # both streams are being read one after another, while stderr is last - it will clear the queue
+            self.streams.clear()
+        return res
+
+
+class OrderedWriter:
+    encoding = sys.getdefaultencoding()
+
+    def __init__(self, fd):
```
Comment (blueyed) [inline src/_pytest/capture.py]:
```suggestion
    def __init__(self, fd: int) -> None:
```
I.e. please add type annotations where you can.

File: src/_pytest/capture.py [hunk 5]
```diff
@@ -692,6 +704,39 @@ def snap(self):
         return res
 
 
+class OrderedCapture(SysCapture):
+    """This class uses deque to keep streams order."""
+    streams = collections.deque()
+
+    def _get_writer_class(self):
+        return OrderedWriter(self.fd)
```
Comment (blueyed) [inline src/_pytest/capture.py]: Returns an object, not a class, doesn't it?

File: src/_pytest/capture.py [hunk 6]
```diff
@@ -692,6 +704,39 @@ def snap(self):
...
+
+class OrderedWriter:
+    encoding = sys.getdefaultencoding()
+
+    def __init__(self, fd):
+        super().__init__()
+        self._fd = fd  # type: int
+
+    def write(self, text, **kwargs):
+        OrderedCapture.streams.append((text, self._fd))
```
Comment (blueyed) [inline src/_pytest/capture.py]: Shouldn't use a class attribute, should it?

File: src/_pytest/capture.py [hunk 7]
```diff
@@ -632,20 +641,23 @@ def snap(self):
...
         if tmpfile is None:
             if name == "stdin":
                 tmpfile = DontReadFromInput()
             else:
-                tmpfile = CaptureIO()
+                tmpfile = self._get_writer_class()
         self.tmpfile = tmpfile
+        
+    def _get_writer_class(self):
+        return CaptureIO()
```
Comment (blueyed) [inline src/_pytest/capture.py]: Instead of this indirection it should get passed in via `tmpfile` really I think.

**Comment by blueyed** [comment]
@aklajnert 
Wouldn't using the same `tmpfile` work? (to have it in order / a single stream)
Of course, the `fd` information is not available then, but that could be provided via a specialed class (your `OrderedCapture`) then, which should be passed in via `tmpfile` then though.

File: src/_pytest/capture.py (hunk 4 above)
Comment (aklajnert) [reply src/_pytest/capture.py]: Sure, I'll do that. I wasn't sure if that's allowed.

File: src/_pytest/capture.py (hunk 5 above)
Comment (aklajnert) [reply src/_pytest/capture.py]: right, the name is wrong

File: src/_pytest/capture.py (hunk 6 above)
Comment (aklajnert) [reply src/_pytest/capture.py]: I did that on purpose, but I'm not sure how it will work with stacking. E.g. when a broader-scoped fixture will use `capsys`, how it will interfere with the one in a test. I'll test it out.

**Comment by aklajnert** [comment]
> @aklajnert
> Wouldn't using the same `tmpfile` work? (to have it in order / a single stream)
> Of course, the `fd` information is not available then, but that could be provided via a specialed class (your `OrderedCapture`) then, which should be passed in via `tmpfile` then though.

@blueyed, I think we want to just add the `combined=True` option, and keep the current feature to extract each stream separately. I don't see a way to preserve it with single `tmpfile`.

File: src/_pytest/capture.py (hunk 6 above)
Comment (aklajnert) [reply src/_pytest/capture.py]: I did some tests, and it seems to work exactly like the current `capsys`. Having the `deque` as a class attribute allows us to just reference it without the need to passing it as an argument.

**Comment by blueyed** [comment]
> single `tmpfile`

The idea is to only use a single `CaptureIO` and pass it in for both stdout and stderr: it should get the combined output then.

Ref: https://github.com/pytest-dev/pytest/pull/6765/files#diff-e4269384a671a462a4824e8b06d4644bR691-R698

**Comment by blueyed** [comment]
FWIW I still think that the feature itself is good (since using a single tmpfile would not allow for distinguishing between which stream was used).

I've also just came across 6f385fb4e - where a non-used "mixed" mode was removed (using a single tmpfile).

File: src/_pytest/capture.py (hunk 7 above)
Comment (aklajnert) [reply src/_pytest/capture.py]: I could do that in the `MultiCapture` class, but that would require additional conditions that will complicate the code. I could also create a new class that inherits `MultiCapture`, but it would have to basically duplicate its `__init__()` which could become a maintenance problem in the future. None of those seems reasonable to me. Do you have any other suggestions for that?

**Comment by aklajnert** [comment]
@blueyed I think I have the code in a quite good shape now (except https://github.com/pytest-dev/pytest/pull/6690/files#r382912377 which I'm not sure how to proceed).  
I've added also a `flush=True` argument to `readouterr()` so it is possible to read streams without flushing. This could be useful when reading a whole combined output and then extract both separately to make sure that a particular message went into the correct stream.  

I need to update the PR description, you think it will be better to just create a fresh PR with a new description, or just continue here?

File: src/_pytest/capture.py [hunk 8]
```diff
@@ -351,11 +351,17 @@ def close(self):
...
             self._capture = None
 
-    def readouterr(self):
+    def readouterr(self, combined=False, flush=True):
         """Read and return the captured output so far, resetting the internal buffer.
 
         :return: captured content as a namedtuple with ``out`` and ``err`` string attributes
         """
+        if combined:
+            return CaptureResult(self._get_combined(), None)
```
Comment (blueyed) [inline src/_pytest/capture.py]: Just a quick thought: couldn't `CaptureResult` gain an optional `combined` attribute/property?
It would still provide `out` and `err´ then, but would also have `combined`.

`combined` could even default to `out + err` then (when not combining).

(I think it's bad to actually have the combined output as `out` here)

Comment (aklajnert) [reply src/_pytest/capture.py]: Yeah, I was thinking about it initially, but since the `CaptureResult` is a `namedtuple` I see 2 potential solutions:  
- It should be changed to regular class with a new attribute or a method, that behaves as tuple to preserve current behavior.
- Instead of returning `CaptureResult` for `combined=True`, just return a single string with the combined output.

I think both approaches are bad - the first one has no added value and may cause confusion, and the second one will make `readouterr()` inconsistent and also potentially confusing.

Maybe adding a new `read()` or `readcombined()` method with a different return type would make sense here?

**Comment by aklajnert** [comment]
@blueyed - could you look at my questions in the open discussions? I'm not really sure how / if should I proceed.  
@nicoddemus - may I also ask you to take a look at it if it makes any sense?

**Comment by bluetech** [comment]
Hi @aklajnert, I'd like to review this PR, might get me a bit to get to it though.

From a cursory look, I think the functionality would be good to have, but I also suspect there is a better way to achieve the goal, however it will take me some time to think "how would I do that".

In the meantime, some comments:

- The code uses some global variables (class variables in `OrderedCapture`) -- code is almost always better without them IMO.

- The proposed API of `combined=True` and returning `(combined, None)` seems not ideal to me -- would either add a new method, or just have `combined=True` only return one value (this can be expressed with type annotation overloads) -- not sure which is better ATM.

- I didn't dig into to what `flush=True` does yet, but is it something that is independent of `combined`? If yes, would be good to have it as a separate (prerequisite?) PR, since smaller PRs are easier to review.

- The implementation uses a deque for combining -- one alternative approach that avoids a shared data structure would be to assign monotonic global IDs to each data written in each SysCapture, and then multiple such can be combined while staying independent. Just a thought, I haven't checked if it would work.

**Comment by aklajnert** [comment]
Thanks for the comments @bluetech.  
I'm leaning towards changing `combine=True` to a separate method, which should be more clear in terms of the return value.
The `flush=True` is not limited to `combine=True`, but it will work only for `capsys` that uses `deque`. I can pull it out into a different branch that originates from this one.  

I'll do another approach for removing the deque from global variables.

**Comment by aklajnert** [comment]
@bluetech I've created a new method to read combined output which returns the single output. Also, the `flush=True` has been extracted into another branch which will wait for this one.  

I didn't find how to make the `deque` non-global without overcomplicating the code though.

**Comment by aklajnert** [comment]
@bluetech @nicoddemus - should I discard this PR, or is there still a chance for it to get merged?

**Comment by nicoddemus** [comment]
> @bluetech @nicoddemus - should I discard this PR, or is there still a chance for it to get merged?

Apologies for the delay. We had some internal issues that we have since then been resolved, and we are in the process of picking up old PRs and clearing up old ones, so it might still take a while to get back to your PR.

Thanks for the contribution and for the patience! Feel free to ping back in a week or so if still nobody posts further comments here. 👍

**Comment by nicoddemus** [comment]
@aklajnert thanks again for submitting this PR, we greatly appreciate it.

Decided to take the time to review now, given the work you have put into this already and the usefulness of the feature. 👍 

As @bluetech commented, it would probably better if we could get away with the global variables and tight coupling between `OrderedCapture` and `OrderedWriter`, but looking at the overall design, I see why it was coded this way: currently `MultiCapture` always demands 3 captures (stdout, stderr, stdin), so we need a way to communicate/order the writes.

I think the code as you wrote it is as good as we can get under the current design. I also think we (the maintainers) should take some time later to refactor things to make them nicer, I'm sure this is possible given that this code has evolved organically a lot since the first time it was written.

Having said that, we just need to make sure the API doesn't prevent us from refactoring this to something we believe is better from the maintenance point of view. This is hard to decide now, given that we don't even have a specific refactoring in mind.

Just as an example of a future API that I can imagine coming up later with a redesign:

```python
def test(capsys):
    capsys.combine()
    out, err = capsys.readouterr()
    # out is combined stream, err is None
```

From an user POV this looks reasonable to me because that's how `subprocess` treats this feature: you pass `stderr=STDOUT` to `Popen`, and then `stdout` is the combined streams while `stderr` is None. This comes from the idea that it might be possible to handle the feature directly in `MultiCapture`, or even by the fixture itself. Again, just an example/idea.

So I propose we mark the `read_combined` method as *experimental*, so users use it but are then aware that the API might change in the future, or even pulled back completely if it causes problems that can't be solved easily. 

So in summary, what I propose:

1. We mark the feature as experimental.
2. Merge it.
3. Work on a redesign/rework on the feature later, without much promises on the existing API.

This of course comes from my belief that this can't be easily accomplished on the current design, but if someone else has an idea on how to implement this cleanly now then by all means let us know. 😁

Thoughts @bluetech @asottile @RonnyPfannschmidt?

**Comment by RonnyPfannschmidt** [comment]
Experimental and go :+1:

File: src/_pytest/capture.py [hunk 9]
```diff
@@ -695,14 +710,65 @@ def __init__(self, fd, tmpfile=None):
...
     "capsysbinary": SysCaptureBinary,
 }
 
 
+class OrderedWriter:
+    encoding = sys.getdefaultencoding()
+
+    def __init__(self, fd: int) -> None:
+        super().__init__()
+        self._fd = fd  # type: int
```
Comment (asottile) [inline src/_pytest/capture.py:751]: I think mypy can infer this value

File: src/_pytest/capture.py [hunk 10]
```diff
@@ -695,14 +710,65 @@ def __init__(self, fd, tmpfile=None):
         self.tmpfile = tmpfile
 
 
+class OrderedCapture(SysCapture):
+    """Capture class that keeps streams in order."""
+
+    streams = collections.deque()  # type: collections.deque
```
Comment (asottile) [inline src/_pytest/capture.py:716]: seems a bit odd for this to be a class member and not an instance member (as written this makes it effectively a global variable).  I'd rather see it as an instance variable

File: src/_pytest/capture.py [hunk 11]
```diff
@@ -380,6 +380,17 @@ def readouterr(self):
         self._captured_err = self.captureclass.EMPTY_BUFFER
         return CaptureResult(captured_out, captured_err)
 
+    def read_combined(self) -> str:
+        """
+        Read captured output with both stdout and stderr streams combined, with preserving
+        the correct order of messages.
+        """
+        if self.captureclass is not OrderedCapture:
+            raise AttributeError("Only capsys is able to combine streams.")
```
Comment (asottile) [inline src/_pytest/capture.py:389]: it is usually a design smell for a superclass to know about a particular subclass, perhaps this method should defer to the captured data's class and the particular subclasses can decide whether they implement this api or not

File: src/_pytest/capture.py (hunk 10 above)
Comment (nicoddemus) [reply src/_pytest/capture.py:716]: Definitely, but see my comment in in https://github.com/pytest-dev/pytest/pull/6690#issuecomment-629804771:

> As @bluetech commented, it would probably better if we could get away with the global variables and tight coupling between OrderedCapture and OrderedWriter, but looking at the overall design, I see why it was coded this way: currently MultiCapture always demands 3 captures (stdout, stderr, stdin), so we need a way to communicate/order the writes.

IOW I don't know how to easily fix this, as @aklajnert also commented.

**Comment by asottile** [comment]
(I think it would also be possible to implement this for capfd, but as I've implied above it's difficult / impossible to preserve exact ordering there in software -- maybe there's an api for fd duping that does the right thing but I haven't come across it yet)

File: src/_pytest/capture.py (hunk 10 above)
Comment (asottile) [reply src/_pytest/capture.py:716]: I think all that's needed is to grow another api on the capture classes which represents combined? if that's possible I think that would also fix the problem below of the backward-class-coupling

Comment (nicoddemus) [reply src/_pytest/capture.py:716]: Hmm but currently who deals with each capture class is `MultiCapture`, which always creates 3 capture instances (`Capture(0)`, `Capture(1)`, and `Capture(2)`).

Or do you mean `MultiCapture` to grow this new API, and forward to the underlying `Capture` classes?

Comment (asottile) [reply src/_pytest/capture.py:716]: oh I see :/ maybe nevermind then

Comment (nicoddemus) [reply src/_pytest/capture.py:716]: Yeah it is a bit frustrating, that's why I think we have to rethink the whole design (in a future PR that is).

File: doc/en/capture.rst [hunk 12]
```diff
@@ -170,3 +170,21 @@ as a context manager, disabling capture inside the ``with`` block:
...
+with both ``stdout`` and ``stderr`` streams combined with preserved chronological order.
+
+.. code-block:: python
+
+    def test_combine(capsys):
+        print("I'm in stdout")
+        print("I'm in stderr", file=sys.stderr)
+        print("Hey, stdout again!")
+
+        output = capsys.read_combined()
```
Comment (nicoddemus) [inline doc/en/capture.rst:188]: Hmm this just occurred to me, to follow with the naming of the other methods (`readouterr`), we should stick to `readcombined`.

**Comment by nicoddemus** [comment]
(If we all agree on the approach I suggested, I can finish up the PR by making the appropriate experimental/docs changes required).

**Comment by bluetech** [comment]
The status on my end is: the typing work and this PR prompted me to look at `capture`. IMO the capture code is quite tricky currently. It is hard to understand *exactly* how it operates. I started refactoring it and got distracted by other things, but I intend to go back to it. During the work I got to understand it better, mostly because whenever I tried to change *anything*, something broke, and then I understood why it does what it does :)

My main point is that I'm not sure that capture is currently in a state where it should be getting new features, instead it should be cleaned up first, to have a solid foundation for new work. Hopefully I'll get to it in a ~month's time.

But I understand that it is not fair to block work on some refactor for an undetermined amount of time, so I'm perfectly fine if another maintainer feels the code is good and the feature is worth having. I'll just rebase my work on top.

**Comment by nicoddemus** [comment]
> My main point is that I'm not sure that capture is currently in a state where it should be getting new features, instead it should be cleaned up first, to have a solid foundation for new work. 

Thanks, that's fair and very reasonable.

However in this specific case, we are taking some measures to at least let us change/remove the `read_combined` function if it turns out to be very hard to implement it in a refactoring, and I also think this functionality is very handy to have in general.

> But I understand that it is not fair to block work on some refactor for an undetermined amount of time, so I'm perfectly fine if another maintainer feels the code is good and the feature is worth having. I'll just rebase my work on top.

I'm not an expert on that code, but it seems it should definitely be possible to redesign it to support the feature. And if turns out to be wrong, we can remove the feature if we mark it as experimental; it might upset some people at that point, but then we can point out that the feature would not even be included in the first place otherwise. 😁

**Comment by asottile** [comment]
I do have additional concern that pytest would be giving the false illusion that stdout and stderr can be deterministically ordered if this feature were integrated -- but I'm fine being overridden on that :)

**Comment by nicoddemus** [comment]
> I do have additional concern that pytest would be giving the false illusion that stdout and stderr can be deterministically ordered if this feature were integrated -- but I'm fine being overridden on that :)

As it stands this works only for `capsys`, which unless I'm missing something we can guarantee, but if it in the future we extend this to `capfd`, then we definitely should at least add a remark on the docs about that. 👍

**Comment by asottile** [comment]
> > I do have additional concern that pytest would be giving the false illusion that stdout and stderr can be deterministically ordered if this feature were integrated -- but I'm fine being overridden on that :)
> 
> As it stands this works only for `capsys`, which unless I'm missing something we can guarantee, but if it in the future we extend this to `capfd`, then we definitely should at least add a remark on the docs about that. 👍

my point is, even though we can guarantee the calls are in order for `write(...)` from the python side this doesn't mean that at runtime they will result in in-order writes to the actual streams

**Comment by nicoddemus** [comment]
> my point is, even though we can guarantee the calls are in order for write(...) from the python side this doesn't mean that at runtime they will result in in-order writes to the actual streams

Ahh sorry yes, I did misunderstand what you said before, thanks for the clarification.

I wonder what sort of `subprocess` guarantees for this feature, though. Do they guarantee, or is some that cannot be easily made work 100% of the time?

**Comment by asottile** [comment]
> > my point is, even though we can guarantee the calls are in order for write(...) from the python side this doesn't mean that at runtime they will result in in-order writes to the actual streams
> 
> Ahh sorry yes, I did misunderstand what you said before, thanks for the clarification.
> 
> I wonder what sort of `subprocess` guarantees for this feature, though. Do they guarantee, or is some that cannot be easily made work 100% of the time?

I've never been able to successfully get a process consuming another processes output on >1 streams to have a consistent ordering without adding significant `sleep` in addition to `flush()` calls :( -- as far as I know it's not possible

but I guess as long as this PR is "know that write() was called in this order" and not really guaranteeing what the actual behaviour will end up being then maybe it's fine?

**Comment by nicoddemus** [comment]
> but I guess as long as this PR is "know that write() was called in this order" and not really guaranteeing what the actual behaviour will end up being then maybe it's fine?

Agree, and I think so IMHO. 😁

**Comment by nicoddemus** [comment]
Just to make it clear though, can we all vote (in this comment) if we should go with the proposal above or not?

👍: We mark the function `readcombined` (renamed, for consistency with `readouterr`) as experimental and merge it. We then, in the future, will strive to refactor the module to remove the warts we currently see on it and possibly implement the feature more elegantly.

👎: We leave this unmerged until the time comes to refactor the module and implement the overall feature in a more elegant manner.

**Comment by aklajnert** [comment]
Thank you, folks, for the feedback. I've spent some time figuring out how to implement this and the final solution came out a bit hacky, so I wouldn't mind postponing merge after the module will be refactored (happy to help with the refactor if you want).  

I wasn't aware that the streams are not synchronized by the OS. I thought that if there is no threading involved, both streams will have a deterministic order. Thanks @asottile for pointing that out. I'll play with it a bit when I'll find some time. I guess that everything should be in the correct order when interpreter is executed with `-u`.

**Comment by nicoddemus** [comment]
Hi everyone,

@bluetech you voted to postpone this until the refactor, you still maintain this position? If so I believe we should close this for now, and retake this once the refactoring is done. 👍 

What do you think?

**Comment by bluetech** [comment]
I think we shouldn't merge the approach taken in this PR, so I'll close this. Just to be clear, the feature itself seems useful, so it's not a rejection of the idea itself.

Thanks @aklajnert for working on this!
